dll/cpl/desk/settings.c

```diff
@@ -842,6 +844,43 @@ SwitchDisplayMode(HWND hwndDlg, PWSTR DeviceName, PSETTINGS_ENTRY seInit, PSETTI
     }
 }
 
+static
+PSETTINGS_ENTRY
+FindBestElement(
+    _In_ PSETTINGS_ENTRY pInitialSettings,
+    _In_ PSETTINGS_ENTRY Current)
+{
+    LONG Penalty, SmallestPenalty = MAXLONG;
+    PSETTINGS_ENTRY pBestEntry = NULL;
+
+    /* Go back to 1st element */
+    while (Current->Blink)
+        Current = Current->Blink;
+
+    /* Go through the list */
+    while (Current)
+    {
+        Penalty = 0x100000 * labs(Current->dmBitsPerPel - pInitialSettings->dmBitsPerPel) +
+                  0x100 * labs(Current->dmPelsWidth - pInitialSettings->dmPelsWidth) +
+                  0x100 * labs(Current->dmPelsHeight - pInitialSettings->dmPelsHeight) +
+                  labs(Current->dmDisplayFrequency - pInitialSettings->dmDisplayFrequency);
+        if (Penalty == 0)
+            return Current;
+
+        if (!pBestEntry || SmallestPenalty > Penalty)
+        {
+            SmallestPenalty = Penalty;
+            pBestEntry = Current;
+        }
```

Comment:
Do we need such complex maths for that?

Reply:
If the perfect choice didn't exist, then it will be the nearest choice.